app/models/assignment.rb

```diff
@@ -1232,23 +1230,61 @@ def zip_automated_test_files(user)
     zip_name = "#{self.short_identifier}-testfiles-#{user.user_name}"
     zip_path = File.join('tmp', zip_name + '.zip')
     FileUtils.rm_rf zip_path
-    files_dir = Pathname.new self.autotest_files_dir
     Zip::File.open(zip_path, create: true) do |zip_file|
-      self.autotest_files.map do |file|
-        path = File.join zip_name, file
-        abs_path = files_dir.join(file)
-        if abs_path.directory?
-          zip_file.mkdir(path)
-        else
-          zip_file.get_output_stream(path) { |f| f.print File.read(abs_path.to_s, mode: 'rb') }
-        end
-      end
+      add_test_files_to_zip(zip_file, zip_name)
     end
     zip_path
   end
 
+  # Writes all of this assignment's automated test files to the +zip_dir+ in +zip_file+. Also writes
+  # the tester settings specified in this assignment's properties and spec file to the json file at
+  # +settings_file_path+ in the +zip_file+.
+  def automated_test_config_to_zip(zip_file, zip_dir, settings_file_path)
+    self.add_test_files_to_zip(zip_file, zip_dir)
+    test_specs_path = self.autotest_settings_file
+    test_specs = File.exist?(test_specs_path) ? JSON.parse(File.open(test_specs_path, &:read)) : {}
+    test_spec_content = strip_ids_from_spec(test_specs)
+    misc_settings = {}
+    misc_settings_attr = [:enable_test, :enable_student_tests, :tokens_per_period, :unlimited_tokens,
+                          :token_start_date, :token_period, :non_regenerating_tokens]
+    misc_settings_attr.each { |f| misc_settings[f] = self.send(f) }
+    zip_file.get_output_stream(settings_file_path) do |f|
+      f.write({ settings: misc_settings, spec_data: test_spec_content }.to_json)
+    end
+  end
+
   private
 
+  def strip_ids_from_spec(test_specs)
```

Comment:
This is duplicating a lot of the work that is being done in `AutomatedTestsController#download_specs` but it much more verbose. You can just copy/reuse that code.

Also, you should not be changing the criterion information here, assigning test groups to the proper criterion in the specs file is handled in `update_test_groups_from_specs` 

Is your branch out of date with master maybe? See the current state here: 
https://github.com/MarkUsProject/Markus/blob/368ece8d80eb868a7232577b0545520fd1ddd287/app/helpers/automated_tests_helper.rb#L41-L43
https://github.com/MarkUsProject/Markus/blob/368ece8d80eb868a7232577b0545520fd1ddd287/app/helpers/automated_tests_helper.rb#L53-L59